docs/proposals/20191017-kubeadm-based-control-plane.md

```diff
@@ -351,41 +426,60 @@ spec:
+
+- In order to determine if a Machine to be rolled out, KCP implements the following:
+    - The infrastructureRef link used by each machine at creation time is stored in annotations at machine level.
+    - The kubeadmConfigSpec used by each machine at creation time is stored in annotations at machine level.
+        - If the annotation is not present (machine is either old or adopted), we won't roll out on any possible changes made in KCP's ClusterConfiguration given that we don't have enough information to make a decision. Users should use KCP.Spec.UpgradeAfter field to force a rollout in this case.
+
+##### Rolling update strategy
+
+Currently KubeadmControlPlane supports only one rollout strategy type the `RollingUpdateStrategyType`. Rolling upgrade strategy's behavior can be modified by using `MaxUnavailable` and `MaxSurge` fields. Both field values can be an absolute number 0 or 1 with following rules:
+
+- If `MaxUnavailable` is set to 0 `MaxSurge` needs to be 1 (default values)
+- If `MaxUnavailable` is set to 1 `MaxSurge` needs to be 0
+
+When `MaxUnavailable` is set to 0 and `MaxSurge` is set to 1 the rollout algorithm is as follows:  
+
   - Find Machines that have an outdated spec
   - If there is a machine requiring rollout
     - Scale up control plane creating a machine with the new spec
     - Scale down control plane by removing one of the machine that needs rollout (the oldest out-of date machine in the failure domain that has the most control-plane machines on it)
 
-- In order to determine if a Machine to be rolled out, KCP implements the following:
-    - The infrastructureRef link used by each machine at creation time is stored in annotations at machine level.
-    - The kubeadmConfigSpec used by each machine at creation time is stored in annotations at machine level.
-        - If the annotation is not present (machine is either old or adopted), we won't roll out on any possible changes made in KCP's ClusterConfiguration given that we don't have enough information to make a decision.
-           Users should use KCP.Spec.UpgradeAfter field to force a rollout in this case.
+Setting `MaxUnavailable` to 1 and `MaxSurge` to 0 is recommended for resource constrained environment like bare-metal, OpenStack or vSphere resource pools, etc. The algorithm is the following:
 
-- The controller should tolerate the manual or automatic removal of a replica during the upgrade process. A replica that fails during the upgrade may block the completion of the upgrade. Removal or other remedial action may be necessary to allow the upgrade to complete.
+  - KubeadmControlPlane verifies that control plane replica count is >= 3
+    - If replica count is less than 3 KubeadmControlPlane fallback to default way of rolling out and try to scale up.
```

Comment:
Does this mean the user's settings would be silently ignored? What if there is no available quota to scale up when the control plane count is 1, `MaxUnavailable` is 1, and `MaxSurge` is 0? Wouldn't it be better to avoid ever getting into this situation in the first place by validating the spec and rejecting this combination of values?

Reply:
@CecileRobertMichon I agree that using spec validation would make more sense than fallback to default rollout here.